modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/DatabaseReaderLazyLoader.java

```diff
@@ -175,8 +208,34 @@ private DatabaseReader get() throws IOException {
     }
 
     @Override
-    public synchronized void close() throws IOException {
+    public void close() throws IOException {
+        closed = true;
```

Comment:
I think there's still race here that can lead to loader not being closed, consider this:
t1 calls `close()` and line 212 is executed
t2 calls `preLookup()`
t1 checks line 213 and doesn't close the loader (`currentUsages == 1`)
t2 checks line 95 `LocalDatabases` and sees loader "closed" 
t2 never calls `postLookup()` and loader is never really closed

We should either relay on single atomic variable for both checks (`Integer.MIN_VALUE` as close state for example) or call `postLookup()` in `else` clause in `LocalDatabases`

Reply:
Something like that should work but it's not tested (it uses negative values as indicator of close state, you have to check result of `preLookup` instead of `isClosed`)
```java
boolean preLookup() {
    return currentUsages.updateAndGet(current -> current < 0 ? current : current + 1) > 0;
}

void postLookup() throws IOException {
    if (currentUsages.updateAndGet(current -> current > 0 ? current - 1 : current + 1) == -1) {
        doClose();
    }
}

@Override
public void close() throws IOException {
    if (currentUsages.updateAndGet(u -> -1 - u) == -1) {
        doClose();
    }
}
```

Reply:
Good point, my forgot to add an else clause to `LocalDatabases#getDatabase(...)` method.
I prefer your approach (only one atomic integer that does all the accounting) over this approach.
Also stress tested it and your approach works correctly.